test/distributed/_composable/test_replicate_with_compiler.py

```diff
@@ -0,0 +1,184 @@
+        super().__init__()
+        self.fc1 = nn.Linear(DIM, DIM)
+        self.fc2 = nn.Linear(DIM, DIM)
+        self.fc3 = nn.Linear(DIM, DIM)
+        self.fc4 = nn.Linear(DIM, DIM)
+        self.use_checkpoint = checkpoint
+
+    def forward(self, x):
+        if self.use_checkpoint:
+            _fc1 = checkpoint(self.fc1, x, use_reentrant(False))
+        else:
+            _fc1 = self.fc1(x)
+        return self.fc4(self.fc3(self.fc2(_fc1)))
+
+
+def compiler_fn(gm):
+    def inner_compiler(gm_, example_inputs_):
+        # This backend is the same as inductor.compile. It's created
+        # for easy debugging purpose.
+        return inductor.compile(gm_, example_inputs_)
+
+    gm = torch.compile(gm, fullgraph=True, backend=inner_compiler)
+    return gm
+
+
+class ReplicateTest(MultiProcessTestCase):
+    @property
+    def world_size(self) -> int:
+        return 2
+
+    def setUp(self) -> None:
+        super().setUp()
+        self._spawn_processes()
+
+    def tearDown(self):
+        super().tearDown()
+        try:
+            os.remove(self.file_name)
+        except OSError:
+            pass
+
+    def _test_compile(
+        self,
+        *,
+        use_gpu: bool,
+        no_sync: bool,
+        setup_func: Optional[Callable] = None,
+    ):
+        backend = "nccl" if use_gpu else "gloo"
+        dist.init_process_group(
+            backend=backend,
+            rank=self.rank,
+            world_size=self.world_size,
+            store=dist.FileStore(self.file_name, self.world_size),
+        )
+        if use_gpu:
+            torch.cuda.set_device(f"cuda:{self.rank}")
+            device = torch.device("cuda")
+        else:
+            device = torch.device("cpu")
+
+        torch._dynamo.config.ddp_python_hook = True
+        torch.manual_seed(123)
+        model = Net().to(device)
+        input = torch.randn([1, DIM], device=device)
+
+        compiled_model = torch.compile(replicate(deepcopy(model)), fullgraph=True)
+        compiled_optim = torch.optim.Adam(compiled_model.parameters())
+        model = replicate(model)
+        optim = torch.optim.Adam(model.parameters())
+
+        if setup_func:
+            setup_func(model, compiled_model)
+
+        # Run multiple iterations so that we could test no_sync
+        for i in range(6):
+            # Settting a different random seed so that if the allreduces are not
+            # executed correctly, the gradients won't be correct compared to the
+            # eager DDP.
+            torch.manual_seed(123 + self.rank + i)
+            input = torch.randn([1, DIM], device=device)
+
+            if no_sync and i % 2 == 0:
+                context = replicate.state(model)._ddp.no_sync()
+            else:
+                context = contextlib.nullcontext()
+            with context:
+                loss = model(input).sum()
+                loss.backward()
+
+            compiled_m = getattr(compiled_model, "_orig_mod", compiled_model)
```

Comment:
confused about this line. if orig_mod doesn't exist, we just use compiled_model? In which cases does this happen?

Reply:
Good point. That seems to be some debug code. I changed it back.